tools/generate-builders.ts

```diff
@@ -48,43 +48,77 @@ interface BuilderExtension {
+  },
+  Onevents: {
+    preValidate: `\n    data.actionMode = data.actionMode || 'sequential';`,
+  },
   Callbackstate: {
-    preValidate: `\r\n    data.type = 'callback';`,
+    preValidate: `\n    data.type = 'callback';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Databasedswitch: {
-    preValidate: `\r\n    data.type = 'switch';`,
+    preValidate: `\n    data.type = 'switch';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Delaystate: {
-    preValidate: `\r\n    data.type = 'delay';`,
+    preValidate: `\n    data.type = 'delay';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Eventbasedswitch: {
-    preValidate: `\r\n    data.type = 'switch';`,
+    preValidate: `\n    data.type = 'switch';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Eventstate: {
-    preValidate: `\r\n    data.type = 'event';`,
+    preValidate: `\n    data.type = 'event';
+                        if (data.exclusive == null) data.exclusive = true;`,
   },
   Foreachstate: {
-    preValidate: `\r\n    data.type = 'foreach';
-                  \r\n    //FIXME https://github.com/serverlessworkflow/sdk-typescript/issues/95
-                  \r\n    data.usedForCompensation = data.usedForCompensation || false;`,
+    preValidate: `\n    data.type = 'foreach';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Injectstate: {
-    preValidate: `\r\n    data.type = 'inject';`,
+    preValidate: `\n    data.type = 'inject';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Operationstate: {
-    preValidate: `\r\n    data.type = 'operation';`,
+    preValidate: `\n    data.type = 'operation';
+                        data.actionMode = data.actionMode || 'sequential';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Parallelstate: {
-    preValidate: `\r\n    data.type = 'parallel';`,
+    preValidate: `\n    data.type = 'parallel';
+                        data.completionType = data.completionType || 'and';
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Subflowstate: {
-    preValidate: `\r\n    data.type = 'subflow';`,
+    preValidate: `\n    data.type = 'subflow';
+                        data.waitForCompletion = data.waitForCompletion || false;
+                        data.usedForCompensation = data.usedForCompensation || false;`,
   },
   Function: {
-    preValidate: `\r\n    data.type =  data.type || 'rest';`,
+    preValidate: `\n    data.type =  data.type || 'rest';`,
   },
   Eventdef: {
-    preValidate: `\r\n    data.kind =  data.kind || 'consumed';`,
+    preValidate: `\n    data.kind =  data.kind || 'consumed';`,
+  },
+  End: {
```

Comment:
default value for end is true, 

```
   "end": {
      "oneOf": [
        {
          "type": "boolean",
          "description": "State end definition",
          "default": true
        },
        {
```

sorry didn't spot this before.

Reply:
I think we are hitting a problem here: if end is defaulted to true, it means the dev will have to explicitly set it to false on every state or any object pointing to `#/definitions/end`

Reply:
@tsurdilo ^

Reply:
something like this will work? 

`if (data == null) data = true;`

(I have not tested this)

Reply:
No, it's one level up, you cannot use the builder for primitive types. [`data` will always exists as an object](https://github.com/serverlessworkflow/sdk-typescript/blob/main/src/lib/builder.ts#L33).

Let's take this example:
```typescript
// ...
.states([
  injectstateBuilder()
    .name('Hello State')
    // ...
    .end(true)
    .build()
])
// ...
```

`end` is not "built" so if you want to set its default value, it must be done at `injectstateBuilder` level, not at `endBuilder` level.

Now let's take an other example with multiple states:
```typescript
// ...
.states([
  injectstateBuilder()
    .name('Hello State')
    // ...
    .transition('End state')
    .build(),
  injectstateBuilder()
    .name('End state')
    // ...
    .end(true)
    .build()
])
// ...
```

In this scenario, if we say that `end` is true by default, `states[0].end` will be set to true. In other words, the resulting object will be:
```json
...
"states": [
  {
    "name": "Hello State",
    ...
    "transition": "End state",
    "end": true
  },
  {
    "name": "End state",
    ...
    "end": true
  }
]
...
```

Which is conflicting and incorrect. 

This is the kind of reasons I though `default` [are tricky and might not be a concern of the SDK](https://github.com/serverlessworkflow/sdk-typescript/issues/99#issuecomment-843233393) (or at least not like this, idk). 

`const` like `type` are a concern because they are always the same (well, they are constants :D), they are defined by the context (or define the context depending on how you see it). If I build an 'injectstate' object, I don't want to have to specify `.type('inject')`, that's infered by the fact that I'm building an 'injectstate'. Or the otherway arround, if I receive a state with the property type 'inject', then I know it's an 'injectstate' for instance.

`default` though are things to be interpreted, they are more volatile. In most cases, it's ok to think "if the value is not specified, then it's the default". But in fact, it's a little more subtle. In the case of `end` for instance, it's "if the key has been specified but have no value, then it means the value is true". In other words, the following examples would have the same output:

```typescript
// ...
.states([
  injectstateBuilder()
    .name('Hello State')
    // ...
    .end() // end key exists but no value specified --> it's true
    .build()
])
// ...
```

```typescript
// ...
.states([
  injectstateBuilder()
    .name('Hello State')
    // ...
    .end(true) // well, here it's obviously true :)
    .build()
])
// ...
```

Reply:
Thanks for the clarification @JBBianchi 

`it must be done at injectstateBuilder level, not at endBuilder level.`

then it can be done, I guess

do you see any issue doing it one level up?

`if(!data.end) data.end = true;`

Reply:
not now, I am just asking, 

we can merge this as it is

Reply:
> 
> 
> Thanks for the clarification @JBBianchi
> 
> `it must be done at injectstateBuilder level, not at endBuilder level.`
> 
> then it can be done, I guess
> 
> do you see any issue doing it one level up?
> 
> `if(!data.end) data.end = true;`

Yes I see an issue, I explained it in my comment ^^'

Reply:
ok I see , well, at the end we might need to code something by hand...